File: sdk/runanywhere-swift/Sources/RunAnywhere/Core/Models/Model/ModelCategory.swift
import Foundation

/// Defines the category/type of a model based on its input/output modality
/// This aligns with FrameworkModality for consistency across the SDK
public enum ModelCategory: String, CaseIterable, Codable, Sendable {
    case language = "language"              // Text-to-text models (LLMs)
    case speechRecognition = "speech-recognition"  // Voice-to-text models (ASR)
    case speechSynthesis = "speech-synthesis"      // Text-to-voice models (TTS)
    case vision = "vision"                  // Image understanding models
    case imageGeneration = "image-generation"      // Text-to-image models
    case multimodal = "multimodal"          // Models that handle multiple modalities
    case audio = "audio"                    // Audio processing (diarization, etc.)

    /// Human-readable display name
    public var displayName: String {
        switch self {
        case .language: return "Language Model"
        case .speechRecognition: return "Speech Recognition"
        case .speechSynthesis: return "Text-to-Speech"
        case .vision: return "Vision Model"
        case .imageGeneration: return "Image Generation"
        case .multimodal: return "Multimodal"
        case .audio: return "Audio Processing"
        }
    }

    /// Icon name for UI display
    public var iconName: String {
        switch self {
        case .language: return "text.bubble"
        case .speechRecognition: return "mic"
        case .speechSynthesis: return "speaker.wave.2"
        case .vision: return "photo.badge.arrow.down"
        case .imageGeneration: return "photo.badge.plus"
        case .multimodal: return "sparkles"
        case .audio: return "waveform"
        }
    }

    /// Maps to the corresponding FrameworkModality
    public var frameworkModality: FrameworkModality {
        switch self {
        case .language: return .textToText
        case .speechRecognition: return .voiceToText
        case .speechSynthesis: return .textToVoice
        case .vision: return .imageToText
        case .imageGeneration: return .textToImage
        case .multimodal: return .multimodal
        case .audio: return .voiceToText  // Audio processing often involves voice-to-text
        }
    }

    /// Initialize from a FrameworkModality
    public init?(from modality: FrameworkModality) {
        switch modality {
        case .textToText:
            self = .language
        case .voiceToText:
            self = .speechRecognition
        case .textToVoice:
            self = .speechSynthesis
        case .imageToText:
            self = .vision
        case .textToImage:
            self = .imageGeneration
        case .multimodal:
            self = .multimodal
        }
    }

    /// Whether this category typically requires context length
    public var requiresContextLength: Bool {
        switch self {
        case .language, .multimodal:
            return true
        case .speechRecognition, .speechSynthesis, .vision, .imageGeneration, .audio:
            return false
        }
    }

    /// Whether this category typically supports thinking/reasoning
    public var supportsThinking: Bool {
        switch self {
        case .language, .multimodal:
            return true
        case .speechRecognition, .speechSynthesis, .vision, .imageGeneration, .audio:
            return false
        }
    }
}

/// Extension to help adapters check compatibility
public extension ModelCategory {
    /// Check if this category is compatible with a framework modality
    func isCompatible(with modality: FrameworkModality) -> Bool {
        switch (self, modality) {
        case (.language, .textToText),
             (.speechRecognition, .voiceToText),
             (.speechSynthesis, .textToVoice),
             (.vision, .imageToText),
             (.imageGeneration, .textToImage),
             (.multimodal, .multimodal):
            return true
        case (.audio, .voiceToText):
            // Audio processing often uses voice-to-text frameworks
            return true
        case (.multimodal, _):
            // Multimodal models can work with any modality
            return true
        default:
            return false
        }
    }

    /// Determine category from a FrameworkModality (non-failable)
    static func from(modality: FrameworkModality) -> ModelCategory {
        switch modality {
        case .textToText: return .language
        case .voiceToText: return .speechRecognition
        case .textToVoice: return .speechSynthesis
        case .imageToText: return .vision
        case .textToImage: return .imageGeneration
        case .multimodal: return .multimodal
        }
    }

    /// Determine category from a framework
    static func from(framework: LLMFramework) -> ModelCategory {
        switch framework {
        case .whisperKit, .openAIWhisper:
            return .speechRecognition
        case .llamaCpp, .mlx, .mlc, .execuTorch, .picoLLM, .foundationModels, .swiftTransformers:
            return .language
        case .coreML, .tensorFlowLite, .onnx, .mediaPipe:
            // These are general frameworks that could be any category
            // Default to multimodal since they can handle various types
            return .multimodal
        }
    }

    /// Determine category from format and frameworks
    static func from(format: ModelFormat, frameworks: [LLMFramework]) -> ModelCategory {
        // First check if we have framework hints
        if let firstFramework = frameworks.first {
            return from(framework: firstFramework)
        }

        // Otherwise guess from format
        switch format {
        case .mlmodel, .mlpackage:
            return .multimodal // Core ML models can be anything
        case .gguf, .ggml, .safetensors, .bin:
            return .language // Usually LLMs
        case .tflite, .onnx:
            return .multimodal // Could be anything
        default:
            return .language // Default to language
        }
    }
}

File: sdk/runanywhere-react-native/src/Core/Models/Model/ModelCategory.ts
/**
 * ModelCategory.ts
 *
 * Defines the category/type of a model based on its input/output modality
 *
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Core/Models/Model/ModelCategory.swift
 */

import { FrameworkModality } from '../Framework/FrameworkModality';
import { LLMFramework } from '../Framework/LLMFramework';
import { ModelFormat } from './ModelFormat';

/**
 * Defines the category/type of a model based on its input/output modality
 */
export enum ModelCategory {
  Language = 'language', // Text-to-text models (LLMs)
  SpeechRecognition = 'speech-recognition', // Voice-to-text models (ASR)
  SpeechSynthesis = 'speech-synthesis', // Text-to-voice models (TTS)
  Vision = 'vision', // Image understanding models
  ImageGeneration = 'image-generation', // Text-to-image models
  Multimodal = 'multimodal', // Models that handle multiple modalities
  Audio = 'audio', // Audio processing (diarization, etc.)
}

/**
 * Human-readable display name
 */
export function getModelCategoryDisplayName(category: ModelCategory): string {
  switch (category) {
    case ModelCategory.Language:
      return 'Language Model';
    case ModelCategory.SpeechRecognition:
      return 'Speech Recognition';
    case ModelCategory.SpeechSynthesis:
      return 'Text-to-Speech';
    case ModelCategory.Vision:
      return 'Vision Model';
    case ModelCategory.ImageGeneration:
      return 'Image Generation';
    case ModelCategory.Multimodal:
      return 'Multimodal';
    case ModelCategory.Audio:
      return 'Audio Processing';
  }
}

/**
 * Maps to the corresponding FrameworkModality
 */
export function getFrameworkModality(
  category: ModelCategory
): FrameworkModality {
  switch (category) {
    case ModelCategory.Language:
      return FrameworkModality.TextToText;
    case ModelCategory.SpeechRecognition:
      return FrameworkModality.VoiceToText;
    case ModelCategory.SpeechSynthesis:
      return FrameworkModality.TextToVoice;
    case ModelCategory.Vision:
      return FrameworkModality.ImageToText;
    case ModelCategory.ImageGeneration:
      return FrameworkModality.TextToImage;
    case ModelCategory.Multimodal:
      return FrameworkModality.Multimodal;
    case ModelCategory.Audio:
      return FrameworkModality.VoiceToText; // Audio processing often involves voice-to-text
  }
}

/**
 * Initialize from a FrameworkModality
 */
export function modelCategoryFromModality(
  modality: FrameworkModality
): ModelCategory | null {
  switch (modality) {
    case FrameworkModality.TextToText:
      return ModelCategory.Language;
    case FrameworkModality.VoiceToText:
      return ModelCategory.SpeechRecognition;
    case FrameworkModality.TextToVoice:
      return ModelCategory.SpeechSynthesis;
    case FrameworkModality.ImageToText:
      return ModelCategory.Vision;
    case FrameworkModality.TextToImage:
      return ModelCategory.ImageGeneration;
    case FrameworkModality.Multimodal:
      return ModelCategory.Multimodal;
  }
}

/**
 * Whether this category typically requires context length
 */
export function requiresContextLength(category: ModelCategory): boolean {
  switch (category) {
    case ModelCategory.Language:
    case ModelCategory.Multimodal:
      return true;
    case ModelCategory.SpeechRecognition:
    case ModelCategory.SpeechSynthesis:
    case ModelCategory.Vision:
    case ModelCategory.ImageGeneration:
    case ModelCategory.Audio:
      return false;
  }
}

/**
 * Whether this category typically supports thinking/reasoning
 */
export function supportsThinking(category: ModelCategory): boolean {
  switch (category) {
    case ModelCategory.Language:
    case ModelCategory.Multimodal:
      return true;
    case ModelCategory.SpeechRecognition:
    case ModelCategory.SpeechSynthesis:
    case ModelCategory.Vision:
    case ModelCategory.ImageGeneration:
    case ModelCategory.Audio:
      return false;
  }
}

/**
 * Determine category from a FrameworkModality (non-failable)
 */
export function modelCategoryFromModalityNonFailable(
  modality: FrameworkModality
): ModelCategory {
  switch (modality) {
    case FrameworkModality.TextToText:
      return ModelCategory.Language;
    case FrameworkModality.VoiceToText:
      return ModelCategory.SpeechRecognition;
    case FrameworkModality.TextToVoice:
      return ModelCategory.SpeechSynthesis;
    case FrameworkModality.ImageToText:
      return ModelCategory.Vision;
    case FrameworkModality.TextToImage:
      return ModelCategory.ImageGeneration;
    case FrameworkModality.Multimodal:
      return ModelCategory.Multimodal;
  }
}

/**
 * Determine category from a framework
 */
export function modelCategoryFromFramework(
  framework: LLMFramework
): ModelCategory {
  switch (framework) {
    case LLMFramework.WhisperKit:
    case LLMFramework.OpenAIWhisper:
      return ModelCategory.SpeechRecognition;
    case LLMFramework.LlamaCpp:
    case LLMFramework.MLX:
    case LLMFramework.MLC:
    case LLMFramework.ExecuTorch:
    case LLMFramework.PicoLLM:
    case LLMFramework.FoundationModels:
    case LLMFramework.SwiftTransformers:
      return ModelCategory.Language;
    case LLMFramework.CoreML:
    case LLMFramework.TensorFlowLite:
    case LLMFramework.ONNX:
    case LLMFramework.MediaPipe:
      // These are general frameworks that could be any category
      // Default to multimodal since they can handle various types
      return ModelCategory.Multimodal;
    case LLMFramework.SystemTTS:
      return ModelCategory.SpeechSynthesis;
    default:
      // For any unknown or new frameworks, default to Multimodal
      return ModelCategory.Multimodal;
  }
}

/**
 * Determine category from format and frameworks
 */
export function modelCategoryFromFormatAndFrameworks(
  format: ModelFormat,
  frameworks: LLMFramework[]
): ModelCategory {
  // First check if we have framework hints
  if (frameworks.length > 0 && frameworks[0] !== undefined) {
    return modelCategoryFromFramework(frameworks[0]);
  }

  // Otherwise guess from format
  switch (format) {
    case ModelFormat.MLModel:
    case ModelFormat.MLPackage:
      return ModelCategory.Multimodal; // Core ML models can be anything
    case ModelFormat.GGUF:
    case ModelFormat.GGML:
    case ModelFormat.SafeTensors:
    case ModelFormat.Bin:
      return ModelCategory.Language; // Usually LLMs
    case ModelFormat.TFLite:
    case ModelFormat.ONNX:
      return ModelCategory.Multimodal; // Could be anything
    default:
      return ModelCategory.Language; // Default to language
  }
}
